crates/gen-host-js/src/lib.rs

```diff
@@ -619,37 +694,82 @@ struct Instantiator<'a> {
 
 impl Instantiator<'_> {
     fn instantiate(&mut self) {
-        uwriteln!(
-            self.src.js,
-            "
-                export async function instantiate(instantiateCore, imports) {{\
-            "
-        );
+        let instantiation = self.gen.opts.instantiation;
+
+        if instantiation {
+            uwriteln!(
+                self.src.js,
+                "
+                    export async function instantiate(compileCore, imports) {{\
+                "
+            );
+        }
+
+        // First do the core Wasm compilations in parallel
+        self.src.js.push_str("const [");
+        let mut first = true;
+        for init in self.component.initializers.iter() {
+            if let GlobalInitializer::InstantiateModule(InstantiateModule::Static(idx, _)) = init {
+                if first {
+                    first = false;
+                } else {
+                    self.src.js.push_str(", ");
+                }
+                self.src.js.push_str(&format!("module{}", idx.as_u32()));
+            }
+        }
+        self.src.js.push_str("] = await Promise.all([");
+        first = true;
+        for init in self.component.initializers.iter() {
+            if let GlobalInitializer::InstantiateModule(InstantiateModule::Static(idx, _)) = init {
+                if first {
+                    first = false;
+                } else {
+                    self.src.js.push_str(", ");
+                }
+                // Get the compiled WebAssembly.Module objects in parallel
+                let name = format!("module{}.wasm", idx.as_u32());
+                if self.gen.opts.instantiation {
+                    uwrite!(self.src.js, "compileCore(\"{name}\")",);
+                } else {
+                    let load_wasm = self.gen.intrinsic(Intrinsic::LoadWasm);
+                    uwrite!(
+                        self.src.js,
+                        "{load_wasm}(new URL('./{name}', import.meta.url))",
+                    );
+                }
+            }
+        }
+        self.src.js.push_str("]);\n");
```

Comment:
I've added a new parallel instantiation here, and also switched `instantiateCore` to `compileCore` so that it can be called in parallel (since import objects otherwise require ordered instantiation). Please let me know if this model works for you?